### ISSUE TITLE
borg check --repair not working

### ISSUE THREAD

**Issue by knutov** [borgbackup/borg#4845]
`borg check --repair`is not working with error: "UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb4 in position 0: invalid start byte".

How to fix it?

```
./borg1.1.10 check -v -p --repair repo
'check --repair' is an experimental feature that might result in data loss.
Type 'YES' if you understand this and want to continue: YES (from BORG_CHECK_I_KNOW_WHAT_I_AM_DOING)
Starting repository check
Checking segments 70.5%
Starting repository index check
Completed repository check, no problems found.
Starting archive consistency check...
Analyzing archive repo-2018-05-31-012111 (1/120)
...
Analyzing archive repo-2019-07-15-024527 (39/120)
Exception while unpacking item metadata [chunk: 001604_b63ed1dca930a6932e280705c80c6d068e8985b8b94797b0e2fa61b081be7fb8]
Local Exception
Traceback (most recent call last):
  File "borg/archiver.py", line 4501, in main
  File "borg/archiver.py", line 4433, in run
  File "borg/archiver.py", line 166, in wrapper
  File "borg/archiver.py", line 334, in do_check
  File "borg/archive.py", line 1225, in check
  File "borg/archive.py", line 1612, in rebuild_refcounts
  File "borg/archive.py", line 1559, in robust_iterator
  File "borg/archive.py", line 1541, in valid_item
  File "borg/archive.py", line 1529, in list_keys_safe
  File "borg/archive.py", line 1529, in <genexpr>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb4 in position 0: invalid start byte

Platform: Linux server1 5.0.0-29-generic #31~18.04.1-Ubuntu SMP Thu Sep 12 18:29:21 UTC 2019 x86_64
Linux: debian buster/sid
Borg: 1.1.10  Python: CPython 3.5.7 msgpack: 0.5.6
PID: 26224  CWD: /home/bak01/borg
sys.argv: ['./borg1.1.10', 'check', '-v', '-p', '--repair', 'repo']
SSH_ORIGINAL_COMMAND: None
```

Repo was made with borg 1.1.9 if it does matter.

**Comment by ThomasWaldmann** [comment]
Strange, I'll have a look ...

**Comment by ThomasWaldmann** [comment]
can you try this patch and rerun borg check?
```
diff --git a/src/borg/archive.py b/src/borg/archive.py
index bfd4ccf9..490f08d7 100644
--- a/src/borg/archive.py
+++ b/src/borg/archive.py
@@ -1526,7 +1526,7 @@ def report(msg, chunk_id, chunk_no):
                 logger.error(msg)
 
             def list_keys_safe(keys):
-                return ', '.join((k.decode() if isinstance(k, bytes) else str(k) for k in keys))
+                return ', '.join((k.decode(errors='replace') if isinstance(k, bytes) else str(k) for k in keys))
 
             def valid_item(obj):
                 if not isinstance(obj, StableDict):
```

**Comment by ThomasWaldmann** [comment]
To explain it: it seems that borg tries to deal with an invalid / corrupted item from the items metadata stream there. When trying to prepare a warning message, it stumbles over the non-utf-8 data when trying to decode it to a str. Code should be able to deal with corrupt data at that place, but in 1.1.10, it is not.

**Comment by knutov** [comment]
I downloaded the source zip from "releases" page of 1.1.10, applied the patch, read https://borgbackup.readthedocs.io/en/stable/installation.html#git-installation and still do not understand how to complie the binary. Can you help with this, please?

**Comment by ThomasWaldmann** [comment]
https://borgbackup.readthedocs.io/en/stable/installation.html#dependencies you need to start reading there.

**Comment by ThomasWaldmann** [comment]
I do not have enough information yet to try to diagnose why it got broken. But maybe we know some more if the warning output code does not crash any more.

**Comment by knutov** [comment]
```
pip install pyinstaller;
pyinstaller --log-level DEBUG -F -n borg-${BORG_ARCH} src/borg/__main__.py;
```
Not obvious, this helps to build binary. I think this instructions are missing in documentation.
Also, binary compiled that way is always says
```
borg1.1.10 --version
borg1.1.10 1.1.11.dev0+gaefe327e.d20191122
```
although I did `git checkout tags/1.1.10`.

Patch helped!

```
Analyzing archive repo-2019-07-15-024527 (39/120)
Did not get expected metadata dict when unpacking item metadata (invalid keys: 117, ���) [chunk: 001604_b63ed1dca930a6932e280705c80c6d068e8985b8b94797b0e2fa61b081be7fb8]
Did not get expected metadata dict when unpacking item metadata (not a dictionary) [chunk: 001604_b63ed1dca930a6932e280705c80c6d068e8985b8b94797b0e2fa61b081be7fb8]
[the same multiple times]
Did not get expected metadata dict when unpacking item metadata (not a dictionary) [chunk: 001604_b63ed1dca930a6932e280705c80c6d068e8985b8b94797b0e2fa61b081be7fb8]
Analyzing archive repo-2019-10-08-024652 (120/120)
Deleting 0 orphaned and 122 superseded objects...
Finished deleting orphaned/superseded objects.
Writing Manifest.
Committing repo (may take a while, due to compact_segments)...
Finished committing repo.
Archive consistency check complete, problems found.
```

Is it possible to understand the reason of repo corruption?

**Comment by ThomasWaldmann** [comment]
IIRC there is a little bit about pyinstaller binary building in the docs and more detailled in the `Vagrantfile`.

The past-1.1.10 version number comes from applying the patch on top of 1.1.10.

It looks like part of an archive's metadata stream is corrupted. As the metadata streams of archives are also deduplicated, this may affect one or many archives, depending on whether they refer to the same corrupt chunk.

I'ld check the hard and software involved, this is often an issue in layers below borg.

**Comment by ThomasWaldmann** [comment]
fixed in 1.1-maint, will be in next release. forward port to master tagged.

"fix" means that borg does not crash there any more and is unrelated to the corruption that caused this (the precise root cause for that is unknown yet, but likely hardware).

**Comment by knutov** [comment]
hardware part seems to be good:

* new disks 
* smart is ok
* memory is ok (and it's ecc)
* server grade hardware
* stable version of zfs
* all zfs checks are ok

There were some cases in the past when planned server reboot for upgrade kernel occurs at the same time some repo was adding new snapshot. Can this combination lead to repo corruptions?

Note, it's zfs, so CoW and every written version of any data supposed to be consistent.

Is there some right way to safely interrupt all running `borg serve` (and not allowing new `borg serve` to start) before `reboot`?
